woocommerce/woocommerce-ios
# Add/Edit/Detail Products: product images horizontal scroll view with readonly/editable empty state

Fixes #1417 

## Description
The main purpose of this implementation is to show the header of the product images in product detail and the product add/edit screen. 
I implemented a reusable table view cell called `ProductImagesHeaderTableViewCell`, really easy to integrate, which exposes two callbacks `onImageSelected()` that returns the image selected, and `onAddImage()` through which you can know if the add image action has been fired. The cell embeds a UICollectionView (first time we use this component in the app).
As you can see from screenshots, cell doesn't fit images #1527 at the moment. This will be made in a separated PR, maybe after implementing the new image caching engine #1432, which can be useful in the implementation.

## Changes
I'm sorry if the changes may seem like a lot, I didn't think that the complexity would have risen so much. A lot of code changes are Xib files and methods moved from a class into another (for example product mocks in tests).
- Implemented `UICollectionViewCell+Helpers` which expose some utils on Collection View Cells
- New image `addImage` in `UIImage+Woo`
- Configured the new cell in `ProductDetailsViewModel`
- Configured the new cell in `ProductFormTableViewModel`
- New cell `ProductImagesHeaderTableViewCell`
- Implemented `ProductImagesCollectionViewDatasource` which manage data source of collection view
- Implemented `ProductImagesViewModel` which is the view model of product images
- Implemented `ProductImagesFlowLayout` which is the custom layout used by the collection view
- New cell `ProductImageCollectionViewCell` + tests
- New cell `AddProductImageCollectionViewCell` + tests
- New cell `ExtendedAddProductImageCollectionViewCell` + tests
- Moved product mocks into a new class

## Testing
1) Enable products (Settings -> Experimental Features -> toggle Products to ON)
2) Select products tab
3) Select a product which contains images
4) You'll be able to see all the images of the product on top.
5) Tap the Edit button in the navbar.
6) You'll be able to see all the images plus the "add image" cell.
7) Return to products list
8) Select a product without images
9) You can see the product detail screen without images on top
10) Tap the Edit button in the nav bar.
11) You'll be able to see on top the "Add a product image" on top
12) Make sure CI pass

## Screenshots

| Product detail with images             |  Product detail with no images |
:-------------------------:|:-------------------------:
![Simulator Screen Shot - iPhone 11 Pro - 2019-11-26 at 15 49 47](https://user-images.githubusercontent.com/495617/69644066-72fa4600-1064-11ea-8f9a-6c70a623b3d8.png) | ![Simulator Screen Shot - iPhone 11 Pro - 2019-11-26 at 15 49 57](https://user-images.githubusercontent.com/495617/69644076-755ca000-1064-11ea-9e9d-90f9daf19df3.png)


| Edit product with images             |  Edit product with no images |
:-------------------------:|:-------------------------:
![Simulator Screen Shot - iPhone 11 Pro - 2019-11-26 at 15 51 04](https://user-images.githubusercontent.com/495617/69644229-a9d05c00-1064-11ea-9e56-3f6aeeb52942.png) | ![Simulator Screen Shot - iPhone 11 Pro - 2019-11-26 at 15 51 39](https://user-images.githubusercontent.com/495617/69644235-ac32b600-1064-11ea-9e78-c5be924fd189.png)


Update release notes:
